---
title: "Using ggplot faceting"
output:
  html_notebook:
    code_folding: show
    highlight: zenburn
    number_sections: no
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**_Please follow the instructions in the README file before starting this tutorial._**

<hr style="border:2px solid gray"> </hr>

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

<hr style="border:2px solid gray"> </hr>

# Loading the libraries

```{r}
library(tidyverse)
```

# Erythromycin resistance

* Erythromycin resistance in CC22's conferred by 2473 bp plasmid encoding *repL* and *ermC*

NC_018969.txt from https://www.ncbi.nlm.nih.gov/pubmed/20943864/

* First lets load the overall genome coverage statistics for each isolate

```{r}
coverages.NC_017763 <- read_tsv('stats.coverage.NC_017763.txt', show_col_types = FALSE)
```

* First lets load the read depths for each isolate mapped to the reference plasmid

```{r}
depths <- 'NC_018969.tsv'

data <- read_tsv(file=depths, show_col_types = FALSE)
data
```

* This data file contains the number of reads mapping at each genome position (POS) of the reference (NC_018969.1). The columns are labelled by the name of the alignment file (*.bam)
* Lets rename the first column, as it begins with a # which can cause problems when specifying columns

```{r}
data %>%
  rename(name='#CHROM')
```
* Lets also pivit the data to produce the long form of the table

```{r}
data %>%
  rename(name='#CHROM') %>%
  pivot_longer(cols=c(-name, -POS), names_to = 'isolate', values_to = 'depth')
```
 
 * And we need to extract out the isolate name from the alignment file paths, we can do this with a combination of te `mutate()` and `str_replace()` functions
 * we can also save the data into a variable called `long_data`

```{r}
long_data <- data %>%
  rename(name='#CHROM') %>%
  pivot_longer(cols=c(-name, -POS), names_to = 'isolate', values_to = 'depth') %>%
  mutate(isolate=str_replace(isolate, 'results/alignments/NC_018969_(.+).bam', '\\1'))
long_data
```

* In order to plot for a single isolate we need to first filter the data, then pass the data into `ggplot()`. Here we plot the plasmid position (`POS`) against the number of reads covering that position (`depth`) using `geom_line()` setting the line width to 2

```{r}
long_data %>%
  filter(isolate == 'SEQ045') %>%
  ggplot(aes(x=POS, y=depth)) + 
    geom_line(size=2) 
```

* lets add two arrows to reprsent the position of the genes on the plasmid, we just prove x,y coords and some style attributes

```{r}
long_data %>%
  filter(isolate == 'SEQ045') %>%
  ggplot(aes(x=POS, y=depth, group=isolate)) + 
    geom_line(size=2) +
    geom_segment(aes(x=1,y=1500,xend=477,yend=1500),size=3,color='blue',arrow=arrow(length=unit(0.4,"cm"))) +
    geom_segment(aes(x=1292,y=1500,xend=2026,yend=1500),size=3,color='red',arrow=arrow(length=unit(0.4,"cm"))) 
```

* Finally we can change the axes labels with `labs()` and set the theme to `theme_bw()`

```{r}
long_data %>%
  filter(isolate == 'SEQ045') %>%
  ggplot(aes(x=POS, y=depth, group=isolate)) + 
    geom_line(size=2) +
    labs(x='LlaG1 gene position', y='Sequence depth') +
    geom_segment(aes(x=1,y=1500,xend=477,yend=1500),size=3,color='blue',arrow=arrow(length=unit(0.4,"cm"))) +
    geom_segment(aes(x=1292,y=1500,xend=2026,yend=1500),size=3,color='red',arrow=arrow(length=unit(0.4,"cm"))) +
    theme_bw()
```

* Note that it is sometimes useful to plot the y axis with a log scale, using the function `scale_y_log10()`

```{r}
long_data %>%
  filter(isolate == 'SEQ045') %>%
  ggplot(aes(x=POS, y=depth, group=isolate)) + 
    geom_line(size=2) +
    scale_y_log10() +
    labs(x='LlaG1 gene position', y='Sequence depth') +
    geom_segment(aes(x=1,y=1500,xend=477,yend=1500),size=3,color='blue',arrow=arrow(length=unit(0.4,"cm"))) +
    geom_segment(aes(x=1292,y=1500,xend=2026,yend=1500),size=3,color='red',arrow=arrow(length=unit(0.4,"cm"))) +
    theme_bw()
```
* if we want to plot the same grph for all isolates we can use `facet_wrap()`. not the removal of the `filter()` function and the inclusion of the `group=isolate` in the `ggplot()` function, then we apply `facet_wrap(~isolate)` to complete the process

```{r fig.height=20, fig.width=20}
long_data %>%
  ggplot(aes(x=POS, y=depth, group=isolate)) + 
    geom_hline(yintercept=0, color="red", size=2) + 
    geom_line(size=2) +
#    scale_y_log10() +
    labs(x='LlaG1 gene position', y='Sequence depth') +
    geom_segment(aes(x=1,y=5000,xend=477,yend=5000),size=3,color='blue',arrow=arrow(length=unit(0.4,"cm"))) +
    geom_segment(aes(x=1292,y=5000,xend=2026,yend=5000),size=3,color='red',arrow=arrow(length=unit(0.4,"cm"))) +
    geom_hline(data = coverages.NC_017763, aes(yintercept = coverage), linetype="dashed", color="darkgreen", size=1) +
    theme_bw() +
    facet_wrap(~isolate)
```

* We can save that final figure to a file with `ggsave()`

```{r}
ggsave('NC_018969-mapping.png', height=20, width=20)
```

<hr style="border:2px solid gray"> </hr>

# Session details

* Generate document version details

```{r}
sessionInfo()
```
